GVRf/Framework/src/org/gearvrf/scene_objects/GVRCylinderSceneObject.java

```diff
@@ -112,10 +327,105 @@ public GVRCylinderSceneObject(GVRContext gvrContext, float bottomRadius,
+            meshes[i] = new GVRMesh(gvrContext);
+        }
+
+        int capVertexNumber = 3 * sliceNumber;
+        int bodyVertexNumber = 4 * sliceNumber * stackNumber;
+        float halfHeight = height / 2.0f;
+
+        // top cap
+        // 3 * numSlices
+        if (topRadius > 0) {
+            vertices = new float[3 * capVertexNumber];
+            normals = new float[3 * capVertexNumber];
+            texCoords = new float[2 * capVertexNumber];
+            indices = new char[capVertexNumber];
+
+            createCap(topRadius, halfHeight, sliceNumber, 1.0f, facingOut);
+
+            meshes[0].setVertices(vertices);
+            meshes[0].setNormals(normals);
+            meshes[0].setTexCoords(texCoords);
+            meshes[0].setTriangles(indices);
+
+            children[0] = new GVRSceneObject(gvrContext,
+                    new FutureWrapper<GVRMesh>(meshes[0]),
+                    futureTextureList.get(0));
+            addChildObject(children[0]);        
+        }
+        
+        // cylinder body
+        // 4 * numSlices * numStacks
+        {
+            int triangleNumber = 6 * sliceNumber * stackNumber;
+    
+            vertices = new float[3 * bodyVertexNumber];
+            normals = new float[3 * bodyVertexNumber];
+            texCoords = new float[2 * triangleNumber];
+            indices = new char[triangleNumber];
+    
+            vertexCount = 0;
+            texCoordCount = 0;
+            indexCount = 0;
+            triangleCount = 0;
+
+            createBody(bottomRadius, topRadius, height, stackNumber, sliceNumber,
+                    facingOut);
+    
+            meshes[1].setVertices(vertices);
+            meshes[1].setNormals(normals);
+            meshes[1].setTexCoords(texCoords);
+            meshes[1].setTriangles(indices);
+    
+            children[1] = new GVRSceneObject(gvrContext,
+                    new FutureWrapper<GVRMesh>(meshes[1]),
+                    futureTextureList.get(1));
+            addChildObject(children[1]);        
+        }
+        
+        // bottom cap
+        // 3 * numSlices
+        if (bottomRadius > 0) {
+            vertices = new float[3 * capVertexNumber];
+            normals = new float[3 * capVertexNumber];
+            texCoords = new float[2 * capVertexNumber];
+            indices = new char[capVertexNumber];
+
+            vertexCount = 0;
+            texCoordCount = 0;
+            indexCount = 0;
+            triangleCount = 0;
+
+            createCap(bottomRadius, -halfHeight, sliceNumber, -1.0f, facingOut);
+
+            meshes[2].setVertices(vertices);
+            meshes[2].setNormals(normals);
+            meshes[2].setTexCoords(texCoords);
+            meshes[2].setTriangles(indices);
+
+            children[2] = new GVRSceneObject(gvrContext,
+                    new FutureWrapper<GVRMesh>(meshes[2]),
+                    futureTextureList.get(2));
+            addChildObject(children[2]);        
+        }
+
```

Comment:
These three blocks of codes look very similar. Can you find a way to reduce the code redundancy?